strawberry/permission.py

```diff
@@ -56,12 +56,25 @@ def has_permission(
             "Permission classes should override has_permission method"
         )
 
+    def resolve_permission_sync(self, source: Any, info: Info, **kwargs: Any) -> None:
+        if self.has_permission(source, info, **kwargs):
+            return None
+        else:
+            return self.on_unauthorized()
+
+    async def resolve_permission_async(
```

Comment:
Splitting into sync and async is a breaking change, I'd like to avoid this. We can think about a different solution :)

Reply:
Yeah, it seemed a bit gross, *very* open to a better solution. As mentioned, might be my lack of Python Syntax knowledge here!

Reply:
@vethan Sorry for that, but could you probably adapt some of the stuff from my implementation on discord? 🤔 
https://discord.com/channels/689806334337482765/741575032257511485/1217820662245101609

If I remember correctly it should work with both async and sync by inspecting whenever any of child permissions are async

Reply:
Also if applied operator is the same as the one in `PermissionHolder` I think you could avoid nesting them too by returning something like
```py
PermissionHolder(operator=..., permissions=[*self.permissions, other])
```

Reply:
Your implementation was great for has_permission, but then there was no way to work through on_unauthorised correctly unfortunately :/  Need to split the logic a bit for that

Reply:
Yeah, `on_unauthorised` is a bit hard to implement because we don't know which permission didn't succeed without checking them again (and we can't even do that)